frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -0,0 +1,843 @@
+	<T as frame_system::Config>::AccountId,
+>>::PositiveImbalance;
+pub type NegativeImbalanceOf<T> = <<T as Config>::Currency as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::NegativeImbalance;
+pub type SignedSubmissionOf<T> =
+	SignedSubmission<<T as frame_system::Config>::AccountId, BalanceOf<T>, CompactOf<T>>;
+
+pub type SubmissionIndicesOf<T> =
+	BoundedBTreeMap<ElectionScore, u32, <T as Config>::SignedMaxSubmissions>;
+
+/// Mask type which pretends to be a set of `SignedSubmissionOf<T>`, while in fact delegating to the
+/// actual implementations in `SignedSubmissionIndices<T>`, `SignedSubmissionsMap<T>`, and
+/// `SignedSubmissionNextIndex<T>`.
+#[cfg_attr(feature = "std", derive(DebugNoBound))]
+pub struct SignedSubmissions<T: Config> {
+	indices: SubmissionIndicesOf<T>,
+	next_idx: u32,
+	insertion_overlay: BTreeMap<u32, SignedSubmissionOf<T>>,
+	deletion_overlay: BTreeSet<u32>,
+}
+
+impl<T: Config> SignedSubmissions<T> {
+	/// Get the signed submissions from storage.
+	pub fn get() -> Self {
+		SignedSubmissions {
+			indices: SignedSubmissionIndices::<T>::get(),
+			next_idx: SignedSubmissionNextIndex::<T>::get(),
+			insertion_overlay: BTreeMap::new(),
+			deletion_overlay: BTreeSet::new(),
+		}
+	}
+
+	/// Put the signed submissions back into storage.
+	pub fn put(mut self) {
+		SignedSubmissionIndices::<T>::put(self.indices);
+		SignedSubmissionNextIndex::<T>::put(self.next_idx);
+		for key in self.deletion_overlay {
+			self.insertion_overlay.remove(&key);
+			SignedSubmissionsMap::<T>::remove(key);
+		}
+		for (key, value) in self.insertion_overlay {
+			SignedSubmissionsMap::<T>::insert(key, value);
+		}
+	}
+
+	/// Get the submission at a particular index.
+	fn get_submission(&self, idx: u32) -> Option<SignedSubmissionOf<T>> {
+		if self.deletion_overlay.contains(&idx) {
+			// Note: can't actually remove the item from the insertion overlay (if present)
+			// because we don't want to use `&mut self` here. There may be some kind of
+			// `RefCell` optimization possible here in the future.
+			None
+		} else {
+			self.insertion_overlay
+				.get(&idx)
+				.cloned()
+				.or_else(|| SignedSubmissionsMap::<T>::try_get(idx).ok())
+		}
+	}
+
+	/// Take the submission at a particular index.
+	///
+	/// Note that this function does not examine the insertion overlay. If that behavior is desired,
+	/// try `self.insetion_overlay.remove(&idx).or_else(|| self.take_submission(idx))`.
+	///
+	/// Note that though this function updates the fields `insertion_overlay` and `deletion_overlay`,
+	/// it does not update the field `indices`. That must be handled separately.
+	fn take_submission(&mut self, idx: u32) -> Option<SignedSubmissionOf<T>> {
+		// We never override an existing index in `self.insertion_overlay`, so if we successfully
+		// removed an item from the insertion overlay, we can be confident that it is not also
+		// present in `SignedSubmissionsMap`.
+		self.insertion_overlay.remove(&idx).or_else(|| {
```

Comment:
what if something is already in storage (from a previous block), and is in the insertion_overlay as well? probably this is unreachable for the current pallet code, but I could write a unit test that would make it happen.

Reply:
Yes, that's unreachable from current pallet code. Agree that you could write a test forcing that scenario, but it's easy enough to predict what would happen: the item would be removed from the overlay but not the underlying map--at least until `finalize_signed_phase`, at which point the entire map is drained. We'd store some extra data in storage until the election ended, but as it will have been removed from `self.indices`, it's inaccessible. I don't think there's a practical attack, there.

I think the thing to guard against is later coders coming in and messing with the underlying state in a way that violates the use instructions ("Just use SignedSubmissions, don't mess with this storage manually"). To that end, e0b3a5267 adds some more debug assertions which are intended to make it for people to manually mess with storage in a way that ruins things.

Reply:
Great, thanks!